(PDS-467) Type sizing updates

caseywilliams opened the pull request:
These updates apply design changes outlined in [this Figma file](https://www.figma.com/file/VOPelwQBGRePUk3sFP3WxP/PDS-Type-Refinements?node-id=72%3A192).

caseywilliams commented on packages/sass-variables/_typography.scss:
```diff
@@ -217,21 +275,41 @@ $puppet-type-code-small-font-weight: $puppet-type-weight-normal !default;
+@mixin puppet-type-code-scaled() {
+  @include puppet-type-code();
+  font-size: #{$puppet-type-code-font-size / $puppet-type-body-font-size}em;
+  line-height: inherit;
```
This is not perfect - this generates the correct size for body text (15px body vs. 16px code), and scales other things in a relatively sane way, but it gets maybe a little weird at smaller sizes, where the designed ratio between body & code sizes is different. This means that the "scaled" size inside some small text is not exactly the same as the dedicated `small` code size we use in the Code component, but I'm not sure how to address that in a better way. The example I modified on the Content page demonstrates all the scaled sizes - happy to adjust this approach if people have advice.

caseywilliams commented on packages/sass-variables/_typography.scss:
```diff
@@ -217,21 +275,41 @@ $puppet-type-code-small-font-weight: $puppet-type-weight-normal !default;
+  overflow-x: auto;
   padding: $puppet-common-spacing-base * 3 $puppet-common-spacing-base * 6;
-  white-space: pre;
+  white-space: pre-wrap;
```
Code blocks seemed to be wrapping in some contexts and not in others - I chose wrap here, but open to changing it of course

caseywilliams commented on packages/sass-variables/_typography.scss:
```diff
@@ -32,56 +32,91 @@ $puppet-type-colors: (
+$puppet-type-hero-line-height: calc(60 / 48) !default;
 $puppet-type-hero-font-family: $puppet-type-font-family-secondary !default;
 $puppet-type-hero-font-weight: $puppet-type-weight-heavy !default;
+$puppet-type-hero-margin: 32px !default;
```
Wasn't totally sure whether to add these margins as variables, but I could imagine using them if I was making a custom component in a separate project, so I went with it.

caseywilliams commented on packages/react-components/source/scss/library/components/_table.scss:
```diff
@@ -50,6 +50,12 @@
   text-align: left;
+
+  code {
+    @include puppet-type-code-small();
```
It seems like it might be a little unusual to generate a table that had a code tag in it, but I updated this since I was already making these changes in the tables withing Content

nmuldavin commented on packages/sass-variables/_typography.scss:
```diff
@@ -263,3 +341,10 @@ $puppet-type-code-small-font-weight: $puppet-type-weight-normal !default;
   }
 }
+
+@mixin puppet-type-link-scaled() {
+  @include puppet-type-link();
+  font-size: inherit;
+  line-height: inherit;
+  font-weight: inherit;
+}
```
I don't think we need this. The `@puppet-type-link` mixin should already work independent of line-height.

vine77 commented:
With respect to the API of non-`Content` components to support nesting (e.g. `<Code size="h2">`), @nmuldavin and I synced again and decided that, it would be nice to support all the body sizes as well as heading sizes and "inherit" (in addition to the others)... However, we can do this in a phased approach with 2 and 3 to be evaluated for need before spending effort on it.

1) `size` props accept "tiny", "small", "regular", "medium", "large" 
2) `size` also accepts "h1" - "h6"
3) `size` also accepts "inherit"

So hopefully just 1 reduces scope and is what we were doing here already.

caseywilliams commented:
Ok, I think I've addressed all the feedback in these revised commits, but let me know if I've gotten something wrong. A few notes:

* There are now consistent "tiny", "small", "regular", "medium", and "large" sizes for Text, Link, and Code (this is phase 1, as described - I haven't tried to add size props for h1-h6 or 'inherit')
* Code and Text are not allowed to scale automatically inside Content. They get the default body size, or the user can pass a specific `size` prop.
* `<a>` tags do scale automatically inside Content because they are styled with the `puppet-type-link` mixin, which does not interfere with size or face (but let me know if I should restrict this more by forcing them to match the body size).
* `<code>` tags inside Content do not scale automatically, since the sizes of the Inconsolata type scale are not consistent with the Open Sans or Calibre type scale. I've applied the different sizes as best I could inside Content, but right now there's no way to get a correctly sized `<code>` tag inside a heading (this would be resolved by doing phase 2 (h1-h6 size props) or phase 3 ('inherit' size prop) as described). You can still see examples of all this on the Content page.
* I did need to update the Table component a little to ensure that the 'small' Link size was used there alongside other 'small' text.

nmuldavin commented:
This is looking great, thank you for being so incredibly thorough. 

The remaining work is to account for the changes to line-height in existing components. The cases I've noticed in particular are the input (maybe other form fields too), and the Alert. Both of those are 4px too tall with the change to small text line height. There are other things whose sizes have changed as a result of this work but we might be ok with that? One example is the items in a select dropdown. It might be worth having a designer @Sigler @jtmuz comb through this to help us identify components that need fixing

caseywilliams commented:
@nmuldavin I tried adding a little padding inside `.rc-input` and `.rc-alert` to account for the line height difference - they at least look vertically centered to me now

nmuldavin commented on packages/react-components/source/scss/library/components/forms/_input.scss:
```diff
@@ -6,7 +6,7 @@
   box-sizing: border-box;
   outline: none;
-  padding: 5px 7px;
+  padding: 7px;
```
I think what you want is to change this to `3px 7px`, to maintain the 32px height with the change from 20px line height to 24px

nmuldavin commented on packages/react-components/source/scss/library/components/_alerts.scss:
```diff
@@ -69,7 +69,7 @@ $colors: (
 .rc-alert-primary-icon {
   flex-grow: 0;
   flex-shrink: 0;
-  margin: $puppet-common-spacing-base * 2;
+  margin: ($puppet-common-spacing-base * 5) / 2;
   margin-right: 0;
 }
```
I think what we want is for this margin to be 8px and then for the padding in `.rc-alert-message` to be `6px 8px` instead of just `8px`

caseywilliams commented:
Thanks @nmuldavin! I've made the changes you describe - the sizing in the preview looks equivalent to the existing styles to me now, except for the intentional changes.

Also: It seems like something may have changed in the github check environment such that errors like [this one](https://github.com/puppetlabs/design-system/pull/253/checks?check_run_id=778488295) started happening for any package using `@babel/preset-env` at v7.8.7. I included just enough version bumps here to get the checks to pass, but let me know if there's more/other things you'd like done related to that.

vine77 commented:
This was released as:
 - @puppet/react-components => 6.0.0-alpha.1
 - @puppet/sass-variables => 2.0.0-alpha.1
